portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -26,7 +34,62 @@
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    // Get all messages stored on Datastore
+    ArrayList<String> messages = new ArrayList<String>();
+    for (Entity entity : results.asIterable()) {
+
+      String comment = (String) entity.getProperty("comment");
+      messages.add(comment);
+    }
```

Comment:
Optional stretch goal: You can do this entire segment (extracting the message text from each comment row) in 1 or 2 lines using Java streams. https://winterbe.com/posts/2014/07/31/java8-stream-tutorial-examples/

This is a more advanced topic so feel free to leave it as-is for now. Streams are widely used inside Google (including in Javascript) so they're a helpful thing to learn at some point this summer

Reply:
Thank you for the resource! I've bookmarked it for now; I'll look into it as soon as I've finished the MVP for this week!